### PR TITLE
Campaign Empire 02: improved scenario continuity

### LINKED ISSUE #3930: Improve "food infrastructure" objective in emp02.wmf scenario
### No objective after food infrastructure objective

Player can get stuck waiting for next objective after building food infrastructure.

- Player is instructed to explore to east and build marble mine in mountains
- When the mine is built, next objective is assigned (build food infrastructure)
- When food infrastructure is built (at least one brewery _and_ at least one tavern OR inn _and_ at least one fishers/hunters house OR piggery OR bakery), the campaign scenario waits until player sees field `38, 33`. After that, "mine infrastructure" objective starts

**Problem:** Nothing motivates him to explore in that direction.
**Proposed solution:** Intermediate objective (e.g. "Explore to the south of marble mountains") which would trigger the "mine infrastructure" objective.
**Alternative:** Drop following code: https://github.com/widelands/widelands/blob/237b030f270b9a6f8a25b3d639cf48cc509051a5/data/campaigns/emp02.wmf/scripting/mission_thread.lua#L106-L108

### Possible deadlock: out of marble columns

Player is instructed to build various food facilities:

> Build a tavern to produce some food to sustain our miners.
> 
> - Of course this means you will need more resources for preparing this food – like fish, meat or bread.
> - To provide these foodstuffs, you can build a fisher’s house, a hunter’s house, a farm, a mill and a bakery.
> 
> Build a brewery and a winery to supply the miners with some good and strong drinks.
> 
> - Maybe you will also need a piggery, a vineyard and some wells.
> - It’s up to you what you want to build. But remember – coal and iron mines need beer, marble and gold mines need wine and all mines need at least rations, which are produced out of bread OR meat OR fish.

**Problem:** Player might try to build all types of buildings and run out of marble columns during the process. Multiple construction sites might get stuck waiting for marble columns.
**Proposed solution:** Allow player to build Stonemason's house.
**Alternative:** Increase initial stock of marble columns.

________________________________________________
**Version:**
 - OS: Debian GNU/Linux 10 (buster)
 - Widelands Version: r24619 (237b030@master)
________________________________________________

Let's discuss if this would be an improvement or not.

Save file: [An Outpost for Exile - deadlock.zip](https://github.com/widelands/widelands/files/4711108/An.Outpost.for.Exile.-.deadlock.zip)

### PR DESCRIPTION
**Pull request by pospec4444**
This fixes both #3930 issues at once.

* Added intermediate "expand south" objective to "marble mine" part
For "No objective after food infrastructure objective" issue.

* Synchronized "marble mine" and "mining/food infrastructure"
For "Possible deadlock: out of marble columns" issue.
If player ignores expansion **and** runs out of columns, he is reminded by new objective "expand south". After expansion, new mining objective starts and Stonemason's house is enabled.

Tested by passive, objective-driven play.

### CODE REVIEW

File: data/campaigns/emp02.wmf/scripting/texts.lua [hunk 1]
```diff
@@ -185,8 +194,11 @@ saledus_2 = {
       _([[It would be a good source for bolstering our supply of quality marble, beyond the meager quantities available from the quarry. Perhaps you were right when you said that this island was like paradise.]])
       .. paragraphdivider() ..
       -- TRANSLATORS: Saledus
-      _([[Please, expand to that mountain and start mining marble.]]))
-      .. new_objectives(obj_build_marblemine)
+      _([[Please, expand to that mountain and start mining marble.]])
+      .. paragraphdivider() ..
+      -- TRANSLATORS: Saledus
+      _([[When you get there, continue with expansion to the south. That area seems also promising.]]))
```
Comment (gunchleoc) [inline data/campaigns/emp02.wmf/scripting/texts.lua]:
```suggestion
      _([[When you get there, continue with expansion to the south. That area also seems promising.]]))
```

File: data/campaigns/emp02.wmf/scripting/texts.lua [hunk 2]
```diff
@@ -56,6 +56,15 @@ obj_build_marblemine = {
    ),
 }
 
+obj_expand_south = {
+   name = "explore_south",
+   title =_"Continue expansion",
+   number = 1,
+   body = objective_text(_"Continue expansion",
+      li(_[[Expand to the south of marble mountain.]])
```
Comment (gunchleoc) [inline data/campaigns/emp02.wmf/scripting/texts.lua]:
```suggestion
      li(_[[Expand farther to the south of the marble mountain.]])
```

File: data/campaigns/emp02.wmf/scripting/texts.lua [hunk 3]
```diff
@@ -56,6 +56,15 @@ obj_build_marblemine = {
    ),
 }
 
+obj_expand_south = {
+   name = "explore_south",
+   title =_"Continue expansion",
+   number = 1,
+   body = objective_text(_"Continue expansion",
```
Comment (gunchleoc) [inline data/campaigns/emp02.wmf/scripting/texts.lua]:
```suggestion
   body = objective_text(_"Continue Expansion",
```

**Review by gunchleoc — CHANGES_REQUESTED** [review]
Player now has to expand south before they can build a stonemason - that needs to be unblocked.

**Comment by pospec4444** [comment]
> Player now has to expand south before they can build a stonemason - that needs to be unblocked.

That's what I tried to address/explain by:

> If player ignores expansion **and** runs out of columns, he is _reminded_ by new objective "expand south". After expansion, new mining objective starts and Stonemason's house is enabled.

Is this OK? https://github.com/widelands/widelands/pull/4038/commits/9dbaa015a54d647f27998940541891bd124e3c76

**Comment by gunchleoc** [comment]
There is an objective that explicitly asks the player to build a stonemason - I still think this should come immediately after the marble mine is finished, because it is teaching the player how to use the marble. Also, if the player should happen to build a stonemason on their own before the food infrastructure is finished, the "Build a stonemason" objective no longer makes sense.

We should put this in a separate function to run in its own thread.

**Comment by pospec4444** [comment]
> There is an objective that explicitly asks the player to build a stonemason

It's not that explicit:
https://github.com/widelands/widelands/blob/d185afc7e6795ba4f3cd9b22fa0fb7c837282b5b/data/campaigns/emp02.wmf/scripting/texts.lua#L59-L66

It is just one part of the objective:
https://github.com/widelands/widelands/blob/d185afc7e6795ba4f3cd9b22fa0fb7c837282b5b/data/campaigns/emp02.wmf/scripting/mission_thread.lua#L136-L161


> I still think this should come immediately after the marble mine is finished, because it is teaching the player how to use the marble.

Like start "mining infrastructure" objective right after marble mine is built?

**Comment by gunchleoc** [comment]
It does say to build a stonemason's house, which is the part I'm interested in.

What I mean is we should have `obj_build_marblemine` done => start `obj_build_mining_infrastructure` and do not wait for expansion south.

**Review by gunchleoc — APPROVED** [review]

**Pull request merged**